src/unix/os390-syscalls.c

```diff
@@ -0,0 +1,372 @@
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+
+#include "os390-syscalls.h"
+#include <errno.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <search.h>
+
+#define CW_CONDVAR 32
+
+#pragma linkage(BPX4CTW, OS)
+#pragma linkage(BPX1CTW, OS)
+
+#define PGTH_CURRENT  1
+#define PGTH_LEN      26
+#define PGTHAPATH     0x20
+#pragma linkage(BPX4GTH, OS)
+#pragma linkage(BPX1GTH, OS)
+
+static int number_of_epolls;
+struct _epoll_list* _global_epoll_list[MAX_EPOLL_INSTANCES];
+
+int alphasort(const void *a, const void *b) {
+
+  return strcoll( (*(const struct dirent **)a)->d_name,
+                  (*(const struct dirent **)b)->d_name );
+}
+
+int scandir(const char *maindir, struct dirent ***namelist,
+            int (*filter)(const struct dirent *),
+            int (*compar)(const struct dirent **,
+            const struct dirent **)) {
+  struct dirent **nl = NULL;
+  struct dirent *dirent;
+  size_t count = 0;
+  size_t allocated = 0;
+  DIR *mdir;
+
+  mdir = opendir(maindir);
+  if (!mdir)
+    return -1;
+
+  while (1) {
+    dirent = readdir(mdir);
+    if (!dirent)
+      break;
+    if (!filter || filter(dirent)) {
+      struct dirent *copy;
+      copy = (struct dirent *)malloc(sizeof(*copy));
+      if (!copy) {
+        while (count) {
+          dirent = nl[--count];
+          free(dirent);
+        }
+        free(nl);
+        closedir(mdir);
+        errno = ENOMEM;
+        return -1;
+      }
+      memcpy(copy, dirent, sizeof(*copy));
+
+      nl = (struct dirent *)realloc(nl, count+1);
+      nl[count++] = copy;
+    }
+  }
+
+  qsort(nl, count, sizeof(struct dirent *),
+      (int (*)(const void *, const void *))compar);
+
+  closedir(mdir);
+
+  *namelist = nl;
+  return count;
+}
+
+static int isfdequal(const void* first, const void* second) {
+  const struct pollfd* a = (const struct pollfd*)first;
+  const struct pollfd* b = (const struct pollfd*)second;
+  return a->fd == b->fd ? 0 : 1;
+}
+
+int epoll_create1(int flags)
```

Comment:
I think you don't need to emulate epoll.  We have different uv__io_t abstractions and you should be able to accomodate a poll(2) based implementation there.  Did you run into trouble?